presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapSslSocketFactory.java

```diff
@@ -0,0 +1,108 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.password.ldap;
+
+import javax.net.SocketFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLSocketFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.InetAddress;
+import java.net.Socket;
+
+public class LdapSslSocketFactory
+        extends SSLSocketFactory
+{
+    private static final ThreadLocal<SSLContext> SSL_CONTEXT = new ThreadLocal<>();
+
+    private final SSLSocketFactory socketFactory;
+
+    public LdapSslSocketFactory(SSLContext sslContext)
+    {
+        this.socketFactory = sslContext.getSocketFactory();
+    }
+
+    @Override
+    public String[] getDefaultCipherSuites()
+    {
+        return socketFactory.getDefaultCipherSuites();
+    }
+
+    @Override
+    public String[] getSupportedCipherSuites()
+    {
+        return socketFactory.getSupportedCipherSuites();
+    }
+
+    @Override
+    public Socket createSocket(Socket socket, String host, int port, boolean autoClose)
+            throws IOException
+    {
+        return socketFactory.createSocket(socket, host, port, autoClose);
+    }
+
+    @Override
+    public Socket createSocket(String host, int port)
+            throws IOException
+    {
+        return socketFactory.createSocket(host, port);
+    }
+
+    @Override
+    public Socket createSocket(String host, int port, InetAddress clientAddress, int clientPort)
+            throws IOException
+    {
+        return socketFactory.createSocket(host, port, clientAddress, clientPort);
+    }
+
+    @Override
+    public Socket createSocket(InetAddress address, int port)
+            throws IOException
+    {
+        return socketFactory.createSocket(address, port);
+    }
+
+    @Override
+    public Socket createSocket(InetAddress address, int port, InetAddress clientAddress, int clientPort)
+            throws IOException
+    {
+        return socketFactory.createSocket(address, port, clientAddress, clientPort);
+    }
+
+    @Override
+    public Socket createSocket(Socket socket, InputStream consumed, boolean autoClose)
+            throws IOException
+    {
+        return socketFactory.createSocket(socket, consumed, autoClose);
+    }
+
+    @Override
+    public Socket createSocket()
+            throws IOException
+    {
+        return socketFactory.createSocket();
+    }
+
+    @SuppressWarnings("MethodOverridesStaticMethodOfSuperclass")
+    public static SocketFactory getDefault()
+    {
+        return new LdapSslSocketFactory(SSL_CONTEXT.get());
+    }
+
+    public static void setSslContext(SSLContext sslContext)
```

Comment:
package private

Reply:
I'd rather leave this public since it's part of the public API for the class. If someone wanted to reuse this class, it'd be useless without it, and there doesn't seem to be any danger of leaving it public.

Reply:
the reason i would want this to be p-private is that it sets a static state (the ThreadLocal, but still a static state)
setting static state should be guarded

if we want to reuse, we will either copy or rename this class, we can revisit